lib/puppet/indirector/request.rb

```diff
@@ -217,11 +217,11 @@ def do_request(srv_service=:puppet, default_server=nil, default_port=nil, &block
       if primary_server = Puppet.settings[:server_list][0]
         bound_port = primary_server[1]
       else
-        bound_port = nil
+        bound_port = Puppet.settings[:masterport]
       end
     end
-    self.server = default_server || bound_server || Puppet.settings[:server]
-    self.port   = default_port || bound_port || Puppet.settings[:masterport]
+    self.server = default_server || bound_server
+    self.port   = default_port || bound_port
```

Comment:
If failover fails, we still set the `[server,port]` pair to `[nil,nil`]. This means that the lookup here will succeed. Moving the `server` lookup inside the rescue means it will only run if we did not attempt failover at all - meaning either `server_list` is unset and we'll fall back to server, or `server_list` is set and we hit a codepath that didn't attempt failover, and we'll use the first element of that array.